waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -264,12 +265,35 @@
                      :revision-timestamp {:pod revision-timestamp :rs rs-revision-timestamp}})
           true))))
 
+;; forward declaration of the hard-delete-service-instance function
+(declare hard-delete-service-instance)
+
+(let [kill-frequently-restarting-pods-thread-pool (Executors/newFixedThreadPool 4)]
+  (defn kill-frequently-restarting-pods
+    "Processes killing of a frequently restarting pod (service instance).
+     On non-leader routers it returns immediately without doing anything.
+     On the leader it triggers killing of the instance asynchronously without adjusting ReplicaSet replica count."
+    [{:keys [leader?-fn] :as scheduler} {:keys [k8s/pod-name k8s/restart-count service-id] :as instance}]
+    (if (leader?-fn)
+      (au/execute
+        (fn kill-frequently-restarting-pods-task []
+          (try
+            (log/info "deleting frequently restarting pod"
+                      {:pod-name pod-name :restart-count restart-count :service-id service-id})
+            (hard-delete-service-instance scheduler instance)
+            (catch Exception ex
+              (log/error ex "error in deleting frequently restarting pod" pod-name))))
+        kill-frequently-restarting-pods-thread-pool)
+      (log/info "skipping deleting frequently restarting pod on non-leader"
+                {:pod-name pod-name :restart-count restart-count :service-id service-id}))))
+
 (defn pod->ServiceInstance
   "Convert a Kubernetes Pod JSON response into a Waiter Service Instance record."
-  [{:keys [api-server-url] :as scheduler} pod]
+  [{:keys [api-server-url restart-kill-threshold] :as scheduler} pod]
   (try
     (let [;; waiter-app is the first container we register
-          primary-container-restart-count (get-in pod [:status :containerStatuses 0 :restartCount] 0)
+          primary-container-restart-count (or (get-in pod [:status :containerStatuses 0 :restartCount]) 0)
+          frequently-restarting? (>= primary-container-restart-count restart-kill-threshold)
```

Comment:
I think `frequently` doesn't match the semantics here since we're not checking frequency (i.e., we don't care how long the pod was live between restarts). Since we only care how many times it restarted, `exceeded-restart-kill-threshold?` seems more accurate.